nltk/tag/perceptron.py

```diff
@@ -45,7 +47,14 @@ def __init__(self):
         # Number of instances seen
         self.i = 0
 
-    def predict(self, features):
+
+    def _softmax(self, scores):
+        s = np.fromiter(scores.values(), dtype=float)
+        exps = np.exp(s)
+        return exps / np.sum(exps)
+
+
+    def predict(self, features, conf=False):
```

Comment:
Explicit is better than implicit, esp. when it comes to arguments that users' can access. 

I would recommend, `def predict(self, features, confidence=False)`

Reply:
Lets use the default `return_conf=True` and `confidence=True`, so that return types can be explicit when used by `tag()` or `train()`.

Reply:
Or in fact make it always return confidence and do away with the mess of allowing multiple return types =)

Reply:
Hang on though -- calling the softmax has a computational cost.  Shouldn't the client be able to decide to not incur that extra cost?

Reply:
Hmmm... Multiple return types is really nasty though. Let me read up on some API design. 

Current ideas:
 1. Multiple return type
 2. Spaghetti function, e.g. `predict_with_conf()` -> `predict()` and add softmax

Reply:
We want to not break current client code, right?

Perhaps a separate function (your #2) is the least disruptive solution.  Not beautiful, but functional?

Reply:
Let sit on this for a day.

I'll look through some C++ design and other Python code design to see whether there's a better solution. It's because this is not the only instance of this so it's better to find a convention that we're are comfortable and let newer code from henceforth follow it =)

Reply:
How about we still return the tuple of (best_label, conf) but use default `conf=False` and put an if-else, e.g.


```python
def predict(self, features, return_conf=False):
    ...
    # Do a secondary alphabetic sort, for stability.
    best_label = max(self.classes, key=lambda label: (scores[label], label))
    # Compute the confidence.
    conf = max(self._softmax(scores)) if return_conf else None

    return best_label, conf
```